File: src/subside/plugins/koth/utils/MessageBuilder.java
package subside.plugins.koth.utils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;

import lombok.Setter;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;

import subside.plugins.koth.KothPlugin;
import subside.plugins.koth.areas.Area;
import subside.plugins.koth.areas.Capable;
import subside.plugins.koth.areas.Koth;
import subside.plugins.koth.captureentities.Capper;
import subside.plugins.koth.gamemodes.TimeObject;
import subside.plugins.koth.modules.KothHandler;
import subside.plugins.koth.scheduler.Schedule;
import subside.plugins.koth.scheduler.ScheduleHandler;

public class MessageBuilder {
    private StrObj message;
    private Collection<Player> excluders;
    private Collection<Player> includers;
    private static @Setter KothPlugin plugin;

    private class StrObj {
        private String[] message;

        private StrObj(String[] message) {
            if (message == null){
                this.message = new String[] {};
            } else {
                this.message = message.clone();
            }
        }

        protected StrObj replaceAll(String search, String replace) {
            for (int x = 0; x < message.length; x++) {
                try {
                    message[x] = message[x].replaceAll(search, replace);
                } catch(Exception e){
                    System.out.println(search+" : "+replace);
                    e.printStackTrace();
                }
            }
            return this;
        }

        protected String[] build() {
            // A way to allow %ttn% to be used everywhere
            if(plugin != null){
                this.replaceAll("%ttn%", TimeObject.getTimeTillNextEvent(plugin));
            }

            for (int x = 0; x < message.length; x++) {
                message[x] = ChatColor.translateAlternateColorCodes('&', message[x]);
            }
            return message;
        }

    }

    public MessageBuilder(String[] msg) {
        this.message = new StrObj(msg);
    }
    
    public MessageBuilder(String msg){
        this.message = new StrObj(new String[]{msg});
    }

    public MessageBuilder koth(KothHandler kothHandler, String koth) {
        // Make sure that $ placeholders aren't causing issues
        koth = koth.replaceAll("\\$", "\\\\\\$");
        
        Koth kth = kothHandler.getKoth(koth);
        if (kth != null) {
            koth(kth);
        } else {
            message.replaceAll("%koth%",
                    koth.replaceAll("\\\\\\\\", "%5C")
                            .replaceAll("([^\\\\])_", "$1 ")
                            .replaceAll("\\\\_", "_")
                            .replaceAll("%5C", "\\\\\\\\"));
        }
        return this;
    }

    public MessageBuilder location(Location location){
        if (location != null) {
            message.replaceAll("%x%", "" + location.getBlockX())
                    .replaceAll("%y%", "" + location.getBlockY())
                    .replaceAll("%z%", "" + location.getBlockZ());
            try {
                message.replaceAll("%world%", location.getWorld().getName());

                if(plugin.getConfigHandler().getGlobal().isWorldFilter()){
                    this.include(location.getWorld().getPlayers());
                }

            } catch (Exception e) { e.printStackTrace(); }
        }
        return this;
    }
    
    public MessageBuilder koth(Koth koth){
        message.replaceAll("%koth%", koth.getName()
                .replaceAll("\\\\\\\\", "%5C")
                .replaceAll("([^\\\\])_", "$1 ")
                .replaceAll("\\\\_", "_")
                .replaceAll("%5C", "\\\\\\\\"));
        this.location(koth.getMiddle());
        return this;
    }
    
    public MessageBuilder area(Area area){
        message.replaceAll("%area%", area.getName());
        this.location(area.getMiddle());
        return this;
    }

    public MessageBuilder area(String area){
        message.replaceAll("%area%", area);
        return this;
    }
    
    public MessageBuilder loot(String loot){
        message.replaceAll("%loot%", loot);
        return this;
    }
    
    public MessageBuilder entry(String entry){
        message.replaceAll("%entry%", entry);
        return this;
    }
    
    public MessageBuilder title(String title){
        message.replaceAll("%title%", title);
        return this;
    }

    public MessageBuilder capper(String capper) {
        message.replaceAll("%capper%", (capper != null) ? capper: "None");
        return this;
    }

    public MessageBuilder day(String day) {
        message.replaceAll("%day%", day);
        return this;
    }

    public MessageBuilder lootAmount(int lootamount) {
        message.replaceAll("%lootamount%", lootamount + "");
        return this;
    }

    public MessageBuilder time(String time) {
        message.replaceAll("%time%", time);
        return this;
    }

    public MessageBuilder times(String times){
        message.replaceAll("%times%", times);
        return this;
    }

    public MessageBuilder captureTime(int captureTime) {
        message.replaceAll("%ct%", "" + captureTime);
        return this;
    }
    
    public MessageBuilder timeTillNext(Schedule schedule){
        message.replaceAll("%ttn%", TimeObject.getTimeTillNextEvent(schedule));
        return this;
    }

    public MessageBuilder id(int id) {
        message.replaceAll("%id%", "" + id);
        return this;
    }

    public MessageBuilder date(String date) {
        message.replaceAll("%date%", date);
        return this;
    }

    public MessageBuilder time(TimeObject tO) {
        message.replaceAll("%m%", String.format("%02d", tO.getMinutesCapped()));
        message.replaceAll("%s%", String.format("%02d", tO.getSecondsCapped()));
        message.replaceAll("%ml%", String.format("%02d", tO.getMinutesLeft()));
        message.replaceAll("%sl%", String.format("%02d", tO.getSecondsLeft()));

        return this;
    }

    public MessageBuilder maxTime(int maxTime) {
        message.replaceAll("%mt%", "" + ((int) maxTime / 60));
        return this;
    }

    public MessageBuilder command(String command) {
        message.replaceAll("%command%", command);
        return this;
    }

    public MessageBuilder commandInfo(String commandInfo) {
        message.replaceAll("%command_info%", commandInfo);
        return this;
    }

    public MessageBuilder exclude(Collection<Player> excluders){
        if(this.excluders == null)
            this.excluders = new ArrayList<>();

        this.excluders.addAll(excluders);
        return this;
    }

    public MessageBuilder include(Collection<Player> includers){
        if(this.includers == null)
            this.includers = new ArrayList<>();

        this.includers.addAll(includers);
        return this;
    }

    public MessageBuilder exclude(Capper<?> capper, Capable area){
        if(capper != null)
            this.exclude(capper.getAvailablePlayers(area));

        return this;
    }

    public void buildAndBroadcast() {
        Collection<? extends Player> players = (includers == null) ? new ArrayList<>(Bukkit.getOnlinePlayers()) : includers;
        
        if(excluders != null){
            players.removeAll(excluders);
        }
        buildAndSend(players);
    }
    
    public void buildAndSend(Capper<?> capper, Capable area){
        if(capper != null)
            buildAndSend(capper.getAvailablePlayers(area));
    }
    
    public void buildAndSend(Collection<? extends Player> players){
        String[] msg = build();
        for(Player player : players){
            for (int x = 0; x < msg.length; x++) {
                if(!msg[x].equalsIgnoreCase("")){
                    Utils.sendMsg(player, msg[x]);
                }
            }
        }
    }
    

    public void buildAndSend(CommandSender player) {
        Utils.sendMsg(player, (Object[])build());
    }

    public String[] build() {
        return message.build();
    }
    
    public List<String> buildArray(){
        return Arrays.asList(message.build());
    }
}

File: src/subside/plugins/koth/utils/MessageBuilder.kt
package subside.plugins.koth.utils

import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.Location
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import subside.plugins.koth.KothManager
import subside.plugins.koth.areas.Area
import subside.plugins.koth.areas.Cappable
import subside.plugins.koth.areas.Koth
import subside.plugins.koth.captureentities.Capper
import subside.plugins.koth.gamemodes.TimeObject
import subside.plugins.koth.modules.KothHandler
import subside.plugins.koth.scheduler.Schedule

class MessageBuilder(message: Array<String>) {
    constructor(message: String): this(arrayOf(message))

    private val message: StrObj = StrObj(message.clone())

    private var excluders: MutableCollection<Player>? = null
    private var includers: MutableCollection<Player>? = null


    private class StrObj(val message: Array<String>) {
        fun replaceAll(search: String, replace: String): StrObj {
            for (i in 0..message.size) {
                try {
                    message[i] = message[i].replace(search, replace)
                } catch (e: Exception) {
                    println("$search : $replace")
                    e.printStackTrace()
                }
            }
            return this
        }

        fun build(): Array<String> {
            for (i in 0..message.size) {
                message[i] = ChatColor.translateAlternateColorCodes('&', message[i])
            }
            return message
        }
    }

    fun koth(kothManager: KothManager, koth: String): MessageBuilder {
        val koth = koth.replace("\\$".toRegex(), "\\\\\\$")
        kothManager.kothHandler.getKoth(koth)
            ?.let { koth(kothManager, it) }
            ?: koth(koth)

        return this
    }

    fun koth(kothManager: KothManager, koth: Koth): MessageBuilder {
        koth(koth.name)
        koth.middle?.let { location(kothManager, it) }

        return this
    }

    fun koth(koth: String): MessageBuilder {
        message.replaceAll(
            "%koth%",
            koth.replace("\\\\\\\\".toRegex(), "%5C")
                .replace("([^\\\\])_".toRegex(), "$1 ")
                .replace("\\\\_".toRegex(), "_")
                .replace("%5C".toRegex(), "\\\\\\\\")
        )

        return this
    }

    fun location(kothManager: KothManager, location: Location): MessageBuilder {
        message.replaceAll("%x%", location.blockX.toString())
            .replaceAll("%y%", location.blockY.toString())
            .replaceAll("%z%", location.blockZ.toString())
            .replaceAll("%world%", location.world.name)

        if (kothManager.configHandler.global.worldFilter) {
            include(location.world.players)
        }

        return this
    }

    fun area(kothManager: KothManager, area: Area): MessageBuilder {
        message.replaceAll("%area%", area.name)
        return location(kothManager, area.middle)
    }

    fun area(area: String): MessageBuilder {
        message.replaceAll("%area%", area)
        return this
    }

    fun loot(loot: String): MessageBuilder {
        message.replaceAll("%loot%", loot)
        return this
    }

    fun entry(entry: String): MessageBuilder {
        message.replaceAll("%entry%", entry)
        return this
    }

    fun title(title: String): MessageBuilder {
        message.replaceAll("%title%", title)
        return this
    }

    fun capper(capper: String?): MessageBuilder {
        message.replaceAll("%capper%", capper ?: "None")
        return this
    }

    fun day(day: String): MessageBuilder {
        message.replaceAll("%day%", day)
        return this
    }

    fun lootAmount(lootAmount: Int): MessageBuilder {
        message.replaceAll("%lootamount%", lootAmount.toString())
        return this
    }

    fun time(time: String): MessageBuilder {
        message.replaceAll("%time%", time)
        return this
    }

    fun times(times: String): MessageBuilder {
        message.replaceAll("%times%", times)
        return this
    }

    fun captureTime(captureTime: Int): MessageBuilder {
        message.replaceAll("%ct%", captureTime.toString())
        return this
    }

    fun timeTillNext(schedule: Schedule): MessageBuilder {
        message.replaceAll("%ttn%", TimeObject.getTimeTillNextEvent(schedule))
        return this
    }

    fun id(id: Int): MessageBuilder {
        message.replaceAll("%id%", id.toString())
        return this
    }

    fun date(date: String): MessageBuilder {
        message.replaceAll("%date%", date)
        return this
    }


    fun time(tO: TimeObject): MessageBuilder {
        message.replaceAll("%m%", String.format("%02d", tO.minutesCapped))
        message.replaceAll("%s%", String.format("%02d", tO.secondsCapped))
        message.replaceAll("%ml%", String.format("%02d", tO.minutesLeft))
        message.replaceAll("%sl%", String.format("%02d", tO.secondsLeft))

        return this
    }

    fun maxTime(maxTime: Int): MessageBuilder {
        message.replaceAll("%mt%", "" + maxTime / 60)
        return this
    }

    fun command(command: String): MessageBuilder {
        message.replaceAll("%command%", command)
        return this
    }

    fun commandInfo(commandInfo: String): MessageBuilder {
        message.replaceAll("%command_info%", commandInfo)
        return this
    }


    fun exclude(excluders: Collection<Player>): MessageBuilder {
        if (this.excluders == null)
            this.excluders = ArrayList()

        this.excluders?.addAll(excluders)
        return this
    }

    fun exclude(capper: Capper<*>?, area: Cappable): MessageBuilder {
        if (capper != null)
            this.exclude(capper.getAvailablePlayers(area))

        return this
    }

    fun include(includers: Collection<Player>): MessageBuilder {
        if (this.includers == null)
            this.includers = ArrayList()

        this.includers?.addAll(includers)
        return this
    }



    fun broadcast() {
        val players: MutableCollection<Player> = includers ?: Bukkit.getOnlinePlayers().toMutableList()

        excluders?.let { players.removeAll(it) }

        send(players)

    }

    fun send(capper: Capper<*>?, area: Cappable) {
        if (capper != null)
            send(capper.getAvailablePlayers(area))
    }

    fun send(players: Collection<Player>) {
        val msgs = build().filter { it != "" }
        if (msgs.isEmpty()) return

        players.forEach { player ->
            Utils.sendMsg(player, msgs)
        }
    }

    fun send(sender: CommandSender) {
        Utils.sendMsg(sender, build())
    }

    fun build() = message.build()
}